### PR TITLE
Unify rack middleware integrations

### LINKED ISSUE #329: Standardize rack middleware
There are problems with mounting grape (and other) apps on Rails apps. This creates two transactions and flip-flops between the two integrations. To make sure only one transaction gets created standardize the rack middleware integration.

### Rack middlewares to refactor

- [x] Create one middleware at the top of the stack, instrument other middleware #159 
- [x] pure-Rack - EventHandler and GenericInstrumentation middleware
  - [x] Add GenericInstrumentation replacement `InstrumentationMiddleware` that is basically the AbstractMiddleware and doesn't set an action name, but people can use it (without the action name going to "unknown" by default) and they will have response body instrumentation that way.
- [x] Rails - #1089
  - [x] #1107 
- [x] Sinatra - #1097
  - [x] Update auto-sinatra hook/integration to also use event handler - #1112
- [x] Hanami
  - [ ] Avoid having to monkeypatch - https://github.com/appsignal/appsignal-ruby/issues/911
  - [x] Apply the same fix as for Sinatra to not have a boot loop for nested app. - #1130
- [x] Grape
  - [x] Have `appsignal/integrations/grape` require `appsignal` - #1122
  - [x] ~Start AppSignal from the integrations file like for Sinatra and Hanami?~ Can create a boot loop...
  - [x] ~Apply the same fix as for Sinatra to not have a boot loop for nested app. #1105~
- [x] Streaming middleware
  - Deprecated in #1140
- [x] Padrino
  - [x] Auto add EventHandler and instrumentation middleware
  - [x] Apply the same fix as for Sinatra to not have a boot loop for nested app. #1105
- [x] ~Webmachine~ - #1141 (technically not middleware)

## Features

- [x] Instrument response bodies: #1099
- [x] Support nested apps
  - [x] EventHandler - #1101 
  - [x] AbstractMiddleware - #1100 
- [x] Add response status code to transaction metadata - #183
  - [x] Fix issue for errors not reporting any response status code. Should it return a 500? - #1125

## Update docs

- [x] Mention the EventHandler for Rack - https://github.com/appsignal/appsignal-docs/pull/1947
- [x] Mention apps need to call `Appsignal.set_error` in Hanami apps to call `handle_exception` if they want to report the error to AppSignal. [See Hanami exception handling docs](https://guides.hanamirb.org/v2.1/actions/exception-handling/). - https://github.com/appsignal/appsignal-docs/pull/1948
- [x] Update Grape docs: https://github.com/appsignal/appsignal-docs/pull/1949
- [x] Update Grape docs to not require `appsignal` manually - https://github.com/appsignal/appsignal-docs/pull/1950
- [x] Update Grape docs to mention including the EventHandler - https://github.com/appsignal/appsignal-docs/pull/1950
- [x] ~Mention how to instrument sinatra without the auto integrations file~ Now that it supported nested apps there's not a lot of reason to document this.
- [x] Update Rack instrumentation page with a default middleware in combination with the EventHandler - see #1139 for more information
  - https://github.com/appsignal/appsignal-docs/pull/1954

## Could unblock

- [x] Avoid Rails middleware injection issue #155 
  - Doesn't seem to fail on newer Rails versions anymore
- [x] Improve Grape validation error reporting #771
- [x] Could fix #289 if we check the `appsignal.transaction` on the request env in padrino
- [x] ~Instrument other middleware #159~
  - I don't know how... we do track the time between the request start and the app handling the request. Everything in between is "some middleware(s) doing things".

### PR DESCRIPTION
**Pull request by julik**
For the moment this is a sketch, @tombruijn would like to hear your thoughts. The approach goes like this:

* There is just one Appsignal transaction across the middleware stack. Spans will be honored if nested integrations are used, but `complete_current!` will be done just once
* Action / method / etc. should be set "inside-out" - we should ensure that the innermost middleware wins when setting those, this is not covered yet but is the intention
* If `filtered_params` were made available on either `request` or in Rack env with ActionDispatch, use them. This to prevent inadvertently leaking unfiltered params via Rack::Request or similar

There are a few bits and bobs for eg `request_id` but we can address those once we are good with the basic approach. TBH in tests I would like to use actual rails/sinatra/rack_test if possible since I found the setup with this amount of mocks fairly brittle - and testing actual Rails controllers is not as complicated (see https://github.com/julik/zip_tricks/blob/next/spec/zip_tricks/rails_streaming_spec.rb#L27)

The resulting "Übermiddleware" has a few indirections but it does seem worth it to be honest.

Closes #329

### CODE REVIEW

File: lib/appsignal/rack/generic_instrumentation.rb [hunk 1]
```diff
@@ -1,41 +1,104 @@
...
+      # a few edge cases even when GenericInstrumentation gets mounted from a Rails app' `config.ru`
+      # there will still be no involuntary disclosure of data that is supposed to get filtered
+      class FilteredParamsWrapper < SimpleDelegator
+        def filtered_params
+          actual_request = __getobj__
+          if actual_request.respond_to?(:filtered_params)
+            actual_request.filtered_params
+          elsif has_rails_filtered_params?(actual_request)
+            # This will delegate to the request itself in case of Rails
+            actual_request.env["action_dispatch.request"].filtered_params
```
Comment (tombruijn) [inline lib/appsignal/rack/generic_instrumentation.rb:26]: I hope we can move these kinds of checks to the middleware of library they're for. That way the instrumentation of each library is together with all the other instrumentation for that library and easier to find back.

See also my comment on: https://github.com/appsignal/appsignal-ruby/issues/329#issuecomment-1964111012

**Comment by tombruijn** [comment]
> There are a few bits and bobs for eg `request_id` but we can address those once we are good with the basic approach. TBH in tests I would like to use actual rails/sinatra/rack_test if possible since I found the setup with this amount of mocks fairly brittle - and testing actual Rails controllers is not as complicated (see https://github.com/julik/zip_tricks/blob/next/spec/zip_tricks/rails_streaming_spec.rb#L27)

There's a history of too much mocking and stubbing in the test suite. From using doubles everywhere for objects from libraries and asserting method calls on gem internals rather than testing the actual results. If possible, I'd like to rid of that as much as possible and have the tests be close as the real world as possible.

For the AppSignal elements, there's an [old ongoing issue](https://github.com/appsignal/appsignal-ruby/issues/252) about it. Every time I update a spec, I rewrite it to the new style, using the [`Appsignal::Transaction#to_h` method.](https://rubydoc.info/gems/appsignal/3.5.6/Appsignal/Transaction#to_h-instance_method) This requires some setup with [start_agent](https://github.com/appsignal/appsignal-ruby/blob/4f907dc48cb77defebe06d65a0089a1aed7d2468/spec/lib/appsignal/rack/sinatra_instrumentation_spec.rb#L24), [keep_transactions](https://github.com/appsignal/appsignal-ruby/blob/4f907dc48cb77defebe06d65a0089a1aed7d2468/spec/lib/appsignal/rack/sinatra_instrumentation_spec.rb#L26), and calling [created_transaction and last_transaction.to_h](https://github.com/appsignal/appsignal-ruby/blob/4f907dc48cb77defebe06d65a0089a1aed7d2468/spec/lib/appsignal/rack/sinatra_instrumentation_spec.rb#L126-L132). Other examples in the test suite should be easy to find by searching for those method names or having a close look at our [testing spec helpers and moneypatches](https://github.com/appsignal/appsignal-ruby/blob/main/spec/support/testing.rb)

File: lib/appsignal/rack/rails_instrumentation.rb [hunk 2]
```diff
@@ -6,59 +6,27 @@ module Appsignal
   # @api private
   module Rack
     class RailsInstrumentation
```
Comment (tombruijn) [inline lib/appsignal/rack/rails_instrumentation.rb:8]: I'm not quite sure what this is doing now, how this would be used, or how it sets errors.

**Comment by tombruijn** [comment]
It's not yet clear to me how these classes would work together. For Rails, I would expect us to add two middlewares to the Rails middleware stack: one at the very front of the stack to report the whole request duration and one to report the action name, errors and other library specific metadata.

For other libraries I assume there would only be one middleware we would add.

If any are nested middlewares they reuse the existing AppSignal transacction to report their data on.

If it's a lot of work to make it work with the request object on the transaction, I wonder if we could make the transaction less dependent (or not dependent at all) on the request object and instead have the middlewares set that meta/sample data directly on the request from the middlewares.

File: lib/appsignal/rack/generic_instrumentation.rb (hunk 1 above)
Comment (julik) [reply lib/appsignal/rack/generic_instrumentation.rb:26]: Yeah this is a side-effect of trying to use the `filtered_params` no-matter-what. If `request` may be re-assigned this is moot (and too clever).

File: lib/appsignal/rack/rails_instrumentation.rb (hunk 2 above)
Comment (julik) [reply lib/appsignal/rack/rails_instrumentation.rb:8]: It is in progress - like I said, it's a sketch for now (because so many things need to get touched)

**Comment by julik** [comment]
> It's not yet clear to me how these classes would work together. For Rails, I would expect us to add two middlewares to the Rails middleware stack: one at the very front of the stack to report the whole request duration and one to report the action name, errors and other library specific metadata.

This is what I am considering as well. What I do suspect is that for all cases _except_ of "just auto-attaching things using the Railtie" people would already have integrations (middleware) configured manually, and it should keep working. "Sandwiching" two Appsignal middleware layers inside of an existing middleware stack is what this PR is about - the reason why it has inheritance and all that is that if you do not (and only "supplement" the already open transaction with data from a particular integration, say) you put the onus on the user to "also not forget to" install the "outer" middleware - which creates the transaction and makes sure it gets completed/sent off. There is also a lot of potential for nesting, some of it can be quite elaborate - I'm not saying everybody does that, but it is quite neat when necessary. I.e.

```
`config.ru - `use WhateverMiddleware; run Raills.application`
   |__>  `middleware.insert(WhateverMiddleware)
       |___> mount SomeRackApp, path: "/subdir"
           |___>  use_middleware(WhateverMiddlewareForSomeRackApp)
                 |___> run Rack::File.new...
```

So saying "there must be just 2 middleware layers, and one of them must be somewhere on the outside, and and and..." seems a bit too strict. Hence why I think allowing both "app nesting" and just having one integration is justified. But for this every integration needs to be able to both create a transaction and set the transaction properties (the `supplement_...`). Since it is not hard to do - that's what I went for.

> If it's a lot of work to make it work with the request object on the transaction, I wonder if we could make the transaction less dependent (or not dependent at all) on the request object and instead have the middlewares set that meta/sample data directly on the request from the middlewares.

For `filtered_params` and all - yes, that would be preferrable.

**Comment by tombruijn** [comment]
> This is what I am considering as well. What I do suspect is that for all cases except of "just auto-attaching things using the Railtie" people would already have integrations (middleware) configured manually, and it should keep working. 

We could auto add it to the very front of the middleware stack by default. And if people have their own middlewares added we can provide instructions on how to remove and then readd AppSignal to the right position.

Something like:
```ruby
Rails.config.application.middlewares.delete Appsignal::Rack::TopRackInstrumentation
Rails.config.application.middleware.unshift Appsignal::Rack::TopRackInstrumentation
```

> Hence why I think allowing both "app nesting" and just having one integration is justified.

I agree. But if the middlewares can handle existing transactions that should not be a major issue, no? I think closing the transaction in the right place would be the most difficult.

I'll be away for a bit after today so someone else from my team will need to take over.

**Comment by tombruijn** [comment]
I'm looking into refactoring the rack middleware like this. Took the first steps in #1089 and #1097. I will try to apply the ideas from this PR in a new PR based on the new way the middlewares work. Still some things to figure out. I will pick this up as part of #329.

**Pull request closed without merging**